exec/java-exec/src/main/java/org/apache/drill/exec/rpc/security/AuthenticationOutcomeListener.java

```diff
@@ -0,0 +1,238 @@
+              true /** the connection will not be backed up at this point */);
+        } else {
+          // success
+          rpcOutcomeListener.success(null, null);
+        }
+      } catch (final Exception e) {
+        rpcOutcomeListener.failed(RpcException.mapException(e));
+      }
+    }
+  }
+
+  @Override
+  public void interrupted(InterruptedException e) {
+    rpcOutcomeListener.interrupted(e);
+  }
+
+  private static class SaslChallengeContext {
+
+    final SaslMessage challenge;
+    final SaslClient saslClient;
+    final UserGroupInformation ugi;
+
+    public SaslChallengeContext(SaslMessage challenge, SaslClient saslClient, UserGroupInformation ugi) {
+      this.challenge = checkNotNull(challenge);
+      this.saslClient = checkNotNull(saslClient);
+      this.ugi = checkNotNull(ugi);
+    }
+  }
+
+  private interface SaslChallengeProcessor {
+
+    /**
+     * Process challenge from server, and return a response.
+     *
+     * Returns null iff SASL exchange is complete and successful.
+     *
+     * @param context challenge context
+     * @return response
+     * @throws Exception
+     */
+    SaslMessage process(SaslChallengeContext context) throws Exception;
+
+  }
+
+  private static class SaslInProgressProcessor implements SaslChallengeProcessor {
+
+    @Override
+    public SaslMessage process(SaslChallengeContext context) throws Exception {
+      final SaslMessage.Builder response = SaslMessage.newBuilder();
+
+      final byte[] responseBytes = evaluateChallenge(context.ugi, context.saslClient,
+          context.challenge.getData().toByteArray());
+
+      final boolean isComplete = context.saslClient.isComplete();
+      logger.trace("Evaluated challenge. Completed? {}.", isComplete);
+      response.setData(responseBytes != null ? ByteString.copyFrom(responseBytes) : ByteString.EMPTY);
+      // if isComplete, the client will get one more response from server
+      response.setStatus(isComplete ? SaslStatus.SASL_SUCCESS : SaslStatus.SASL_IN_PROGRESS);
+      return response.build();
+    }
+  }
+
+  private static class SaslSuccessProcessor implements SaslChallengeProcessor {
+
+    @Override
+    public SaslMessage process(SaslChallengeContext context) throws Exception {
+      if (context.saslClient.isComplete()) {
+        logger.trace("Successfully authenticated to server using {}", context.saslClient.getMechanismName());
+        // setup security layers here..
+        return null;
+      } else {
+
+        // server completed before client; so try once, fail otherwise
+        evaluateChallenge(context.ugi, context.saslClient,
+            context.challenge.getData().toByteArray()); // discard response
+
+        if (context.saslClient.isComplete()) {
+          logger.trace("Successfully authenticated to server using {}", context.saslClient.getMechanismName());
+          // setup security layers here..
+          return null;
```

Comment:
Looks like DRILL-4335 is requesting for TLS. But encryption via SASL is also possible using [wrap and unwrap](https://docs.oracle.com/javase/7/docs/api/javax/security/sasl/SaslServer.html#unwrap(byte[],%20int,%20int)) functions, if "quality of protection" is "auth-conf" (confidentiality). The differences are [nicely noted here](http://stackoverflow.com/questions/11347304/security-authentication-ssl-vs-sasl).

I do not see any open tickets for encryption using SASL, but that should be a sub-task of [DRILL-291](https://issues.apache.org/jira/browse/DRILL-291).